node/core/approval-voting/src/aux_schema/mod.rs

```diff
@@ -0,0 +1,597 @@
+
+			Some((a.cert().clone(), a.validator_index()))
+		})
+	}
+
+	// Produce a bitvec indicating the assignments of all validators up to and
+	// including `tranche`.
+	pub(crate) fn assignments_up_to(&self, tranche: DelayTranche) -> BitVec<BitOrderLsb0, u8> {
+		self.tranches.iter()
+			.take_while(|e| e.tranche <= tranche)
+			.fold(bitvec::bitvec![BitOrderLsb0, u8; 0; self.assignments.len()], |mut a, e| {
+				for &(v, _) in &e.assignments {
+					a.set(v as _, true);
+				}
+
+				a
+			})
+	}
+}
+
+/// Metadata regarding approval of a particular candidate.
+#[derive(Debug, Clone, Encode, Decode, PartialEq)]
+pub(crate) struct CandidateEntry {
+	pub candidate: CandidateReceipt,
+	pub session: SessionIndex,
+	// Assignments are based on blocks, so we need to track assignments separately
+	// based on the block we are looking at.
+	pub block_assignments: BTreeMap<Hash, ApprovalEntry>,
+	pub approvals: BitVec<BitOrderLsb0, u8>,
+}
+
+impl CandidateEntry {
+	/// Access the bit-vec of approvals.
+	pub(crate) fn approvals(&self) -> &BitSlice<BitOrderLsb0, u8> {
+		&self.approvals
+	}
+}
+
+/// Metadata regarding approval of a particular block, by way of approval of the
+/// candidates contained within it.
+#[derive(Debug, Clone, Encode, Decode, PartialEq)]
+pub(crate) struct BlockEntry {
+	pub block_hash: Hash,
+	pub session: SessionIndex,
+	pub slot: SlotNumber,
+	pub relay_vrf_story: RelayVRFStory,
+	// The candidates included as-of this block and the index of the core they are
+	// leaving. Sorted ascending by core index.
+	pub candidates: Vec<(CoreIndex, CandidateHash)>,
+	// A bitfield where the i'th bit corresponds to the i'th candidate in `candidates`.
+	// The i'th bit is `true` iff the candidate has been approved in the context of this
+	// block. The block can be considered approved if the bitfield has all bits set to `true`.
+	pub approved_bitfield: BitVec<BitOrderLsb0, u8>,
+	pub children: Vec<Hash>,
+}
+
+/// A range from earliest..last block number stored within the DB.
+#[derive(Debug, Clone, Encode, Decode, PartialEq)]
+pub(crate) struct StoredBlockRange(BlockNumber, BlockNumber);
+
+/// Canonicalize some particular block, pruning everything before it and
+/// pruning any competing branches at the same height.
+pub(crate) fn canonicalize(
+	store: &impl AuxStore,
+	canon_number: BlockNumber,
+	canon_hash: Hash,
+)
+	-> sp_blockchain::Result<()>
+{
+	let range = match load_stored_blocks(store)? {
+		None => return Ok(()),
+		Some(range) => if range.0 >= canon_number {
+			return Ok(())
+		} else {
+			range
+		},
+	};
+
+	let mut deleted_height_keys = Vec::new();
+	let mut deleted_block_keys = Vec::new();
+
+	// Storing all candidates in memory is potentially heavy, but should be fine
+	// as long as finality doesn't stall for a long while. We could optimize this
+	// by keeping only the metadata about which blocks reference each candidate.
+	let mut visited_candidates = HashMap::new();
```

Comment:
I suppose we do not have the practice of using `with_capacity` when creating large data structures?  Or we just don't worry since the cost is only logarithmic and maybe annoying to compute?

Reply:
I tend to just use `new` until I am convinced that something needs optimizing. Allocating a few hundred elements should be really cheap, comparitively.

Reply:
It'd not super easy to compute here so never mind.